app/Http/Controllers/CaasController.php

```diff
@@ -3,70 +3,168 @@
     {
-        $request->validate([
-            'shift' => 'required|exists:shifts,id',
-        ]);
-        $selectedShift = Shift::find($request->shift);
-        $plotting = new Plotting([
+        $shift = Shift::where('id', $request->id)->first();
+        $day = Carbon::parse($shift->day)->locale('id')->translatedFormat('l d M');
+        //kalo ada yg iseng masuk lewat link
+        if ($shift->quota <= 0 || Auth::guard('caas')->user()->announcecheck()->isPlotActive == 0) {
+            return redirect()->route('caas.schedule');
+        }
+        $data = [
+            'title' => "Konfirmasi Jadwal",
+            'schedule' => $shift,
+            'day' => $day
+        ];
+        return view('caas.confirmschedule', $data);
+    }
+    public function saveSchedule(Request $request)
+    {
+        $selectedShift = Shift::find($request->id);
+        $quota = $selectedShift->quota - 1;
+        Plotting::create([
             'shift_id' => $selectedShift->id,
+            'caas_id' => Auth::guard('caas')->user()->id,
+            'isPlotActive' => 1,
+            'created_at' => Carbon::now(),
+            'updated_at' => Carbon::now()
+        ]);
+        Shift::where('id', $request->id)->update([
+            'quota' => $quota
         ]);
-        auth()->user()->datacaas()->plots()->save($plotting);
-        return redirect()->route('shift.dipilih')->with('success', 'Shift berhasil dipilih!');
+        return redirect()->route('caas.fix.schedule');
     }
     public function fixSchedule()
     {
-        $selectedShift = auth()->user()->datacaas->plots->shift;
-        return view('shift.dipilih', compact('selectedShift'));
+        // dd(boolval(Auth::guard('caas')->user()->plots));
+        if (boolval(Auth::guard('caas')->user()->plots) == 0) {
+            return redirect()->route('caas.schedule');
+        }
+        $sday = Auth::guard('caas')->user()->plots->shifts->day;
+        $stime = Auth::guard('caas')->user()->plots->shifts->start_hour;
+        $etime = Auth::guard('caas')->user()->plots->shifts->end_hour;
+        $day = Carbon::parse($sday)->locale('id')->translatedFormat('l');
+        $date = Carbon::parse($sday)->locale('id')->translatedFormat('d M, Y');
+        $time = substr($stime, 0, -3) . " - " . substr($etime, 0, -3);
+        $data = [
+            'title' => "Jadwal Terpilih",
+            'day' => $day,
+            'date' => $date,
+            'time' => $time
+        ];
+        return view('caas.fixschedule', $data);
     }
+
+
+    //Role functions
     public function viewRole()
     {
-        $selectRoles = Grouprole::all();
-        return view('roles.caas', compact('selectRoles'));
+        //just validasi kalo ada yg iseng
+        if (Auth::guard('caas')->user()->status->isPass == 0 || Auth::guard('caas')->user()->role->roles_id != 1 || Auth::guard('caas')->user()->announcecheck()->isRoleActive == 0) {
+            return redirect()->route('caas.fix.role');
+        }
```

Comment:
```suggestion
        if (!Auth::guard('caas')->user()->status->isPass || Auth::guard('caas')->user()->role->roles_id != 1 || !Auth::guard('caas')->user()->announcecheck()->isRoleActive) {
            return redirect()->route('caas.fix.role');
        }
```

Actually, these validation should have existed in the routes/web.php, are there any reason why you guys didnt put it there instead? @DiyArc-350

Reply:
I didn't find a better way to make this caas validation more efficient, but I tried to reduce the code length by joining the table inside the model, thus i have to validate the caas guard frequently, also the route for both caas and admin are already protected.

Reply:
i see, so the point here is that i think you havent realized that there is this thing called middleware https://laravel.com/docs/10.x/middleware, you can use that instead so for every single route in which you need to do the custom check and validation, you can just reuse that middleware, can you use this instead?

Reply:
this is an important concept in backend development, so it might as well serve as something that you can practice too 👍

Reply:
will do bang

Reply:
done bang @fakhrip